### PR TITLE
10982 10909 Allow using OAI-PMH identifiers as persistent ids of harvested datasets 

### LINKED ISSUE #10982: Feature Request: Request identifier support for OAI_DC harvesting
**Overview of the Feature Request**

When harvesting in OAI_DC, it is possible that the identifier compatible with Dataverse is not entered in the metadata part.
It is possible in this case that it is in the header and in the `request` tag of the response, example with this DOI:

```
<?xml version="1.0" encoding="UTF-8"?>
<OAI-PMH
  xmlns="http://www.openarchives.org/OAI/2.0/"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/
  http://www.openarchives.org/OAI/2.0/OAI-PMH.xsd">
  <responseDate>2024-10-30T14:31:17.227Z</responseDate>
  <request
    verb="GetRecord"
    identifier="doi:10.13144/lil-1266"
    metadataPrefix="oai_dc">https://data.progedo.fr/oai</request>
  <GetRecord>
    <record>
      <header>
        <identifier>doi:10.13144/lil-1266</identifier>
        <datestamp>2024-10-30</datestamp>
      </header>
      <metadata>
        <oai_dc:dc
          xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/"
          xmlns:dc="http://purl.org/dc/elements/1.1/"
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd">
          <dc:title>International Social Survey Programme - Enquête France - Relations sociales III - 2017</dc:title>
          <dc:identifier>https://data.progedo.fr/codebooks/doi/10.13144/lil-1266</dc:identifier>
        </oai_dc:dc>
      </metadata>
    </record>
  </GetRecord>
</OAI-PMH>
```

The purpose of this issue is to add the identifier of the `request` tag to the list of identifiers retrieved in the metadata part in order to increase the harvesting success rate.

**Any open or closed issues related to this feature request?**
#10909 

**Are you thinking about creating a pull request for this feature?**  
A PR is planned

### PR DESCRIPTION
**Pull request by landreev**
**What this PR does / why we need it**:

This PR reconciles the 2 implementations of the same feature, mine and @stevenferey 's, hence 2 issue numbers in the PR name. It will only close one of them, #10982; the other issue needs more work on other features being added as part of it and it's going back into "on hold" for now. 

**Which issue(s) this PR closes**:

- Closes #10982

**Special notes for your reviewer**:

**Suggestions on how to test this**:

Below is a real life example of an OAI-PMH archive subject to the feature added in this PR. 

Create a harvesting client as follows: 

harvest url: https://data.progedo.fr/oai
set: None
metadata format: oai_dc
archive type: Generic OAI archive

Most (or all?) of their records will be failing to import when testing with the develop branch prior to this PR. All (or most) should successfully import when testing this branch. As of writing this, having tested just now: 

Screen Shot 2024-11-25 at 10 20 33 AM

since this is a real, active archive, their holdings are subject to ongoing change; so, your results may vary. 

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:

**Additional documentation**:

Preview docs at https://dataverse-guide--11049.org.readthedocs.build/en/11049/api/native-api.html#create-a-harvesting-client

### CODE REVIEW

File: src/main/resources/db/migration/V6.4.0.1.sql [hunk 1]
```diff
@@ -1,4 +1,4 @@
 -- Adding a case-insensitive index related to #11003
 --
 
-CREATE UNIQUE INDEX IF NOT EXISTS INDEX_DVOBJECT_authority_protocol_upper_identifier ON dvobject (authority, protocol, UPPER(identifier));
\ No newline at end of file
+CREATE UNIQUE INDEX IF NOT EXISTS INDEX_DVOBJECT_authority_protocol_upper_identifier ON dvobject (authority, protocol, UPPER(identifier));
```
Comment (pdurbin) [inline src/main/resources/db/migration/V6.4.0.1.sql:4]: Will removing the newline make Flyway calculate a different checksum and refuse to deploy this? Should we `git checkout` (revert) the file back to how it was?

File: doc/release-notes/11049-oai-identifiers-as-pids.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+## When harvesting, Dataverse can now use the identifier from the OAI-PMH record header as the persistent id for the harvested dataset.
+
+This will allow harvesting from sources that do not include a persistent id in their oai_dc metadata records, but use valid dois or handles as the OAI-PMH record header identifiers. 
+
+It is also possible to optionally configure a harvesting client to use this OAI-PMH identifier as the **preferred** choice for the persistent id. See the [Harvesting Clients API](https://guides.dataverse.org/en/6.5/api/native-api.html#create-a-harvesting-client) section of the Guides for more information.
```
Comment (pdurbin) [inline doc/release-notes/11049-oai-identifiers-as-pids.md:5]:
```suggestion
It is also possible to optionally configure a harvesting client to use this OAI-PMH identifier as the **preferred** choice for the persistent id. See the [Harvesting Clients API](https://guides.dataverse.org/en/6.5/api/native-api.html#create-a-harvesting-client) section of the Guides, #11049 and #10982 for more information.
```

File: doc/sphinx-guides/source/api/native-api.rst [hunk 3]
```diff
@@ -5280,6 +5281,7 @@ The following optional fields are supported:
...
 - customHeaders: This can be used to configure this client with a specific HTTP header that will be added to every OAI request. This is to accommodate a use case where the remote server requires this header to supply some form of a token in order to offer some content not available to other clients. See the example below. Multiple headers can be supplied separated by `\\n` - actual "backslash" and "n" characters, not a single "new line" character. 
 - allowHarvestingMissingCVV: Flag to allow datasets to be harvested with Controlled Vocabulary Values that existed in the originating Dataverse Project but are not in the harvesting Dataverse Project. (Default is false). Currently only settable using API.
+- useOaiIdentifiersAsPids: Defaults to false; if set to true, Harvester will attempt to use the identifier from the OAI-PMH record header as the **first choice** for the persistent id of the harvested dataset. When set to false, Dataverse will still attempt to use this identifier, but only if none of the `<dc:identifier>` entries in the OAI_DC record contain a valid persistent id (this is new as of v6.5).
```
Comment (pdurbin) [inline doc/sphinx-guides/source/api/native-api.rst:5284]:
```suggestion
- useOaiIdentifiersAsPids: Defaults to false; if set to true, the harvester will attempt to use the identifier from the OAI-PMH record header as the **first choice** for the persistent id of the harvested dataset. When set to false, Dataverse will still attempt to use this identifier, but only if none of the `<dc:identifier>` entries in the OAI_DC record contain a valid persistent id (this is new as of v6.5). 
```

File: src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvestingClient.java [hunk 4]
```diff
@@ -214,6 +214,7 @@ public void setArchiveDescription(String archiveDescription) {
         this.archiveDescription = archiveDescription; 
     }
     
+    @Column(columnDefinition="TEXT")
     private String harvestingSet;
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvestingClient.java:218]: Hmm, what was this before? 255 characters? Now it's unlimited?

File: src/test/java/edu/harvard/iq/dataverse/api/imports/ImportGenericServiceBeanTest.java [hunk 5]
```diff
@@ -1,21 +1,70 @@
...
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.InjectMocks;
 import org.mockito.junit.jupiter.MockitoExtension;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNull;
 
+import java.nio.charset.StandardCharsets;
+
 @ExtendWith(MockitoExtension.class)
 public class ImportGenericServiceBeanTest {
```
Comment (pdurbin) [inline src/test/java/edu/harvard/iq/dataverse/api/imports/ImportGenericServiceBeanTest.java:21]: I'll just put this comment here. API tests didn't run: https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-11049/3/testReport/

**Review by pdurbin — COMMENTED** [review]
Overall, looks great. Here's some initial feedback.

File: src/main/resources/db/migration/V6.4.0.1.sql (hunk 1 above)
Comment (landreev) [reply src/main/resources/db/migration/V6.4.0.1.sql:4]: Yes, good call! 
It did fail to deploy for me this morning, and I just deleted the flyway history entry and deployed again without thinking about it much. I must have introduced this when resolving a conflict with the cherry-picked commit from my other branch.

File: src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvestingClient.java (hunk 4 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvestingClient.java:218]: Yes. The accompanying flyway script adjusts the existing tables accordingly.

Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvestingClient.java:218]: Oh, another good call! That's something that came cherry-picked from my initial branch, and it is NOT needed in this branch. Will remove. 
(we want it to be unlimited, to be able to harvest from DataCite, where we'll be passing potentially very long lists of DOIs as "sets"; but it is not needed in this PR just yet).

File: src/test/java/edu/harvard/iq/dataverse/api/imports/ImportGenericServiceBeanTest.java (hunk 5 above)
Comment (landreev) [reply src/test/java/edu/harvard/iq/dataverse/api/imports/ImportGenericServiceBeanTest.java:21]: The build my latest changes have triggered is still running. I'll keep an eye on it.

Comment (landreev) [reply src/test/java/edu/harvard/iq/dataverse/api/imports/ImportGenericServiceBeanTest.java:21]: The last Jenkins run did succeed.

File: src/main/resources/db/migration/V6.4.0.1.sql (hunk 1 above)
Comment (landreev) [reply src/main/resources/db/migration/V6.4.0.1.sql]: Interestingly, it actually looks like Flyway is smart enough to see that it's still the same script, despite the extra newline - according to a quick experiment. It must be stripping newlines (and maybe extra whitespaces and comments?) before taking the checksum - ? 
my failure to deploy this morning appeared to be related to something else.

File: src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvestingClient.java (hunk 4 above)
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/harvest/client/HarvestingClient.java]: Removed in 8a361be. Thanks.

**Comment by ofahimIQSS** [comment]
tested and passed in internal - 1670 Files were successfully harvested
[Testing of 11049.docx](https://github.com/user-attachments/files/17936332/Testing.of.11049.docx)

**Pull request merged**